posthog/temporal/data_imports/pipelines/pipeline_non_dlt.py

```diff
@@ -0,0 +1,392 @@
+
+        if len(buffer) > 0:
+            py_table = pa.Table.from_pylist(buffer)
+            self._process_pa_table(pa_table=py_table, index=chunk_index)
+            row_count += py_table.num_rows
+
+        self._post_run_operations(row_count=row_count)
+
+    def _process_pa_table(self, pa_table: pa.Table, index: int):
+        delta_table = self._delta_table_helper.get_delta_table()
+
+        pa_table = _append_debug_column_to_pyarrows_table(pa_table, self._load_id)
+        pa_table = _evolve_pyarrow_schema(pa_table, delta_table.schema() if delta_table is not None else None)
+
+        table_primary_keys = self._get_primary_keys()
+        delta_table = self._delta_table_helper.write_to_deltalake(
+            pa_table, self._is_incremental, index, table_primary_keys
+        )
+
+        self._internal_schema.add_pyarrow_table(pa_table)
+
+        _update_incrementality(self._schema, pa_table, self._logger)
+        _update_job_row_count(self._job.id, pa_table.num_rows, self._logger)
+
+    def _post_run_operations(self, row_count: int):
+        delta_table = self._delta_table_helper.get_delta_table()
+
+        assert delta_table is not None
+
+        self._logger.info("Compacting delta table")
+        delta_table.optimize.compact()
+        delta_table.vacuum(retention_hours=24, enforce_retention_duration=False, dry_run=False)
+
+        file_uris = delta_table.file_uris()
+        self._logger.info(f"Preparing S3 files - total parquet files: {len(file_uris)}")
+        prepare_s3_files_for_querying(self._job.folder_path(), self._resource_name, file_uris)
+
+        self._logger.debug("Validating schema and updating table")
+
+        validate_schema_and_update_table_sync(
+            run_id=str(self._job.id),
+            team_id=self._job.team_id,
+            schema_id=self._schema.id,
+            table_schema={},
+            table_schema_dict=self._internal_schema.to_hogql_types(),
+            row_count=row_count,
+            table_format=DataWarehouseTable.TableFormat.DeltaS3Wrapper,
+        )
+
+    def _get_primary_keys(self) -> list[Any] | None:
+        primary_keys = self._resource._hints.get("primary_key")
+
+        if primary_keys is None:
+            return None
+
+        if isinstance(primary_keys, list):
+            return primary_keys
+
+        if isinstance(primary_keys, Sequence):
+            return list(primary_keys)
+
+        raise Exception(f"primary_keys of type {primary_keys.__class__.__name__} are not supported")
+
+
+def _evolve_pyarrow_schema(table: pa.Table, delta_schema: deltalake.Schema | None) -> pa.Table:
+    py_table_field_names = table.schema.names
+
+    # Change pa.structs to JSON string
+    for column_name in table.column_names:
+        column = table.column(column_name)
+        if pa.types.is_struct(column.type) or pa.types.is_list(column.type):
+            json_column = pa.array([json.dumps(row.as_py()) if row.as_py() is not None else None for row in column])
```

Comment:
Out of scope here but an issue I just discovered that might be addressable here. clickhouse s3 can't deserialize a list like `["test"]`

Reply:
Do you more context to this? A support ticket maybe?